### PR TITLE
Add authentication for Safekeeper

### PR DESCRIPTION
**Pull request by yeputons**
Based on #2730

### CODE REVIEW

File: control_plane/src/compute.rs
Comment (yeputons) [inline control_plane/src/compute.rs:310]: This env variable is already set when running the compute node, it's used inside `pageserver_connstring`

**Comment by petuhovskiy** [comment]
Would be good to update https://github.com/neondatabase/neon/blob/main/docs/authentication.md, to follow current state of things better.

IIUC, we currently have 3 storage services operating with Neon JWT:
- compute
- safekeeper
- pageserver

I'd like to have documentation on what tokens are placed in the config of these services, and whether they're optional or not.

In safekeeper and pageserver we have libpq and http interfaces, and we accept queries with tokens on these interfaces. Clarification on what tokens we should use for these interfaces will be also nice.

This is not strictly related to this PR, but you seem to have context for these things which is needed to write a good doc :)
This doc will also make reviewing this PR a little easier.

File: libs/utils/src/auth.rs
Comment (arssher) [inline libs/utils/src/auth.rs:26]: Why All?

File: libs/utils/src/http/endpoint.rs
Comment (arssher) [inline libs/utils/src/http/endpoint.rs:140]: E can/should be std::error:Error

File: control_plane/src/compute.rs
Comment (arssher) [inline control_plane/src/compute.rs:328]: This won't be needed if we make $ZENITH_AUTH_TOKEN the default GUC value -- and better reuse it as a token to both  pageserver and safekeeper (can be in separate pr).

File: control_plane/src/storage.rs
Comment (arssher) [inline control_plane/src/storage.rs:243]: worth a comment 'generate a token to connect from pageserver to safekeeper'

File: safekeeper/src/wal_service.rs
Comment (arssher) [inline safekeeper/src/wal_service.rs:21]: Let's move public key inside `SafekeeperConf`.

**Review by arssher — APPROVED** [review]

File: libs/utils/src/http/endpoint.rs
Comment (yeputons) [reply libs/utils/src/http/endpoint.rs:140]: We use `anyhow::Error` instead. It does not implement `std::error::Error` for some reason. Changed to an explicit type.

File: safekeeper/src/wal_service.rs
Comment (yeputons) [reply safekeeper/src/wal_service.rs:21]: Will do together with corresponding Pageserver refactoring in a separate PR.

**Comment by yeputons** [comment]
@arssher @petuhovskiy Would you mind checking if the updated `authentication.md` makes sense? It looks very similar to a file describing data flow in Neon, if we have one.

File: docs/authentication.md
Comment (arssher) [inline docs/authentication.md:130]: Let mention here config param which specifies public key to check tokens with.

File: docs/authentication.md
Comment (arssher) [inline docs/authentication.md:64]: Word 'for' is redundant here at least.

File: docs/authentication.md
Comment (arssher) [inline docs/authentication.md:94]: I think this is out of scope here, but for example when compute finished earlier than pageserver consumed all WAL stream and it needs to catch up.

File: docs/authentication.md
Comment (arssher) [inline docs/authentication.md:96]: They are not different from main pageserver WRT auth.

File: docs/authentication.md
Comment (petuhovskiy) [inline docs/authentication.md:106]: Typo?

```suggestion
either of two values:
```

**Comment by petuhovskiy** [comment]
Great doc, thanks @yeputons! Will review the code after the rebase, but it looks mostly ok.

File: pageserver/src/bin/pageserver.rs
Comment (petuhovskiy) [inline pageserver/src/bin/pageserver.rs:261]: Random thought, can we rename `ZENITH_AUTH_TOKEN` to `NEON_AUTH_TOKEN`? We don't use it extensively yet, should be easy to rename in all places.

File: libs/utils/src/auth.rs
Comment (yeputons) [reply libs/utils/src/auth.rs:31]: Yes, at the moment, adding a comment.

**Comment by yeputons** [comment]
Follow-up PRs and TODOs:

* [ ] `ZENITH_AUTH_TOKEN` --> `NEON_AUTH_TOKEN`
* [ ] Move move public key inside `SafekeeperConf`/`PageserverConf` and load it on start instead of passing a separate argument through chains of calls (`auth_validation_public_key_path` --> `auth_validation_public_key`).
* [ ] `Claims`: make it more structured. E.g. make `Scope::Tenant, tenant_id=None` impossible.
* [ ] Think about generating tokens for debug purposes when SSHed to a staging machine. E.g. `curl http://localhost:12345/some-pageserver-api` is useful

File: pageserver/src/auth.rs
Comment (yeputons) [reply pageserver/src/auth.rs:11]: I'd prefer a separate PR: merge `Claims`+`Scope` into a single `enum` so it's impossible to construct incorrect data. Does that work?

File: control_plane/src/compute.rs
Comment (yeputons) [reply control_plane/src/compute.rs:326]: Finally decided to keep `NULL` the default as the current default is no-auth.

File: pageserver/src/auth.rs
Comment (petuhovskiy) [reply pageserver/src/auth.rs:11]: Should work, thanks!

**Comment by yeputons** [comment]
Waiting for #2863 to be reviewed, approved, and merged before this.

**Comment by yeputons** [comment]
Rebased on top of #2863 (which is to be merged as soon as tests pass), no big conflicts with either `main` or the other PR.

Will merge this right PR after #2863, if there are no objections.

**Pull request merged**
